### PR TITLE
Added gpg api to add keys

### PR DESCRIPTION
**Pull request by sliverc**
Simply importing gpg keys from keyserver or armored public key.

### CODE REVIEW

File: api/gpg.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+	}
+	if len(b.GpgKeyID) > 0 {
+		args = append(args, "--recv", b.GpgKeyID)
+	}
+
+	cmd := exec.Command("gpg", args...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
```
Comment (winged) [inline api/gpg.go]: Shouldn't we capture stderr/stdout and log/return the output? This way, it might get lost..

File: api/gpg.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		args = append(args, "--import", keypath)
+
+	}
+	if len(b.GpgKeyID) > 0 {
+		args = append(args, "--recv", b.GpgKeyID)
+	}
+
+	cmd := exec.Command("gpg", args...)
+	cmd.Stdin = os.Stdin
```
Comment (winged) [inline api/gpg.go]: I would rather close stdin of the gpg process. Gpg may sometimes try to interact, depending on the configuration. Not useful when we run a server.

File: api/gpg.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			c.Fail(400, err)
+			return
+		}
+		if _, err = keyfile.WriteString(b.GpgKeyArmor); err != nil {
+			c.Fail(400, err)
+		}
+		args = append(args, "--import", keypath)
+
+	}
+	if len(b.GpgKeyID) > 0 {
```
Comment (winged) [inline api/gpg.go:56]: I think this should be an `else if` - What happens if a client specifies both GpgKeyID and GpgKeyArmor?

File: api/gpg.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		Keyring     string
+	}
+
+	if !c.Bind(&b) {
+		return
+	}
+
+	var err error
+	args := []string{"--no-default-keyring"}
+	keyring := "trustedkeys.gpg"
```
Comment (winged) [inline api/gpg.go:27]: I take it this is relative to the .gnupg dir anyway? otherwise, I think some path sanitising would be in order

**Review by winged — CHANGES_REQUESTED** [review]

File: api/gpg.go (hunk 4 above)
Comment (sliverc) [reply api/gpg.go:27]: yes it is...

man page gpg:
```
Add  file to the current list of keyrings. If file begins with a tilde and a slash, these are replaced
by the $HOME directory. If the filename does not contain a slash, it is assumed to  be  in  the  GnuPG
home directory ("~/.gnupg" if --homedir or $GNUPGHOME is not used).
```

File: api/gpg.go (hunk 3 above)
Comment (sliverc) [reply api/gpg.go:56]: if both are defined gpg will raise an error and such error will be returned as 400... I thought this to be safe enough - gpg knows what is allowed and what is not.

Comment (winged) [reply api/gpg.go:56]: In that case, I would add a comment explaining that this is intentional

File: api/gpg.go (hunk 1 above)
Comment (sliverc) [reply api/gpg.go]: go already does this for us.

https://golang.org/pkg/os/exec/#Cmd.Run
https://golang.org/pkg/os/exec/#ExitError

File: api/gpg.go (hunk 3 above)
Comment (sliverc) [reply api/gpg.go:56]: added comment where gpg command is run how error handling is working

**Comment by sliverc** [comment]
@winged 
as discussed also removed not needed stdin and stderr redirection. ready for review again.

**Pull request merged**